Implemented more functionality according to AHM TCP Protocol V1.4

JonasArnold opened the pull request:
**I added the following functionality to the module:**
- Distinguish Type of Device (AHM-64/AHM-32/AHM-16) in order to reduce traffic when using device other than AHM-64
- Action "Increment/Decrement Level of Input"
- Action "Increment/Decrement Level of Zone"
- Action "Playback Track"
- Action "Set Level of Input"
- Action "Set Level of Zone"
- Variables for Levels of Inputs and Zones

**Also I made some changes to the code structure:**
- cleaned up code a bit and 
- tried resolving terminology confusion on the terms "input", "zone", "channel"
- added separate file for constants that are related to the whole module
- added separate file for helper functions that are related more to Companion than the module itself  

**All functionality has been tested using a brand-new AHM-32.** 

**What is still missing:**
- Up version of the module
- Upgrade Scripts for existing Configurations (I do not know if this is required, maybe can update and test what is broken?)
- Testing with AHM-16 and AHM-64 (I have no access)

JonasArnold commented:
Anybody got time to review this and merge this pull request?
 @bryce-seifert 
 @JeffreyDavidsz 

thanks in advance!

bryce-seifert commented:
@JonasArnold I can take a look and will get back to you!

bryce-seifert commented:
@JonasArnold Just looking at the code, not actually testing yet, to answer your Q we will need upgrade scripts anywhere an action's ID is changed. So from a glance, any action affected by the change of `id: 'inputChannel'`, to `id: 'mute_number'` will need it. We'll also need an upgrade to handle the new `ahm_type` config option, just to set a default value for pre-existing installations.

fenderwall commented:
hi, when can we expect this update to come through? We would like to use some of these features too. Thanks!

bryce-seifert commented:
This PR is waiting for changes to be made in order to properly upgrade the actions that were changed, so they don't break existing setups

fenderwall commented:
Ah okay, so @JonasArnold has to make these changes? Thanks

bryce-seifert commented on index.js:
```diff
@@ -52,6 +81,18 @@ class AHMInstance extends InstanceBase {
 			},
+			{
+				type: 'dropdown',
+				id: 'ahm_type',
```
Needs corresponding upgrade script

bryce-seifert commented on actions.js:
```diff
@@ -1,38 +1,32 @@
 				type: 'dropdown',
 				label: name,
-				id: 'inputChannel',
+				id: 'inputNum',
```
Needs corresponding upgrade script

JonasArnold commented:
> @JonasArnold Just looking at the code, not actually testing yet, to answer your Q we will need upgrade scripts anywhere an action option's ID is changed. So from a glance, any action affected by the change of `id: 'inputChannel'`, to `id: 'mute_number'` will need it. We'll also need an upgrade to handle the new `ahm_type` config option, just to set a default value for pre-existing installations.

I had a look at it. I have checked in a first version of the update script  #[d22dc02](https://github.com/bitfocus/companion-module-allenheath-ahm/pull/15/commits/d22dc024538a1b6ee5caa620c7bb3193f45602a9). unfortunately the update of the actions does not work yet. 
Unfortunately I have no way to debug it and also the `console.log`s from the update script do not show in the Companion UI. Can anyone help me with this?

fenderwall commented:
@JonasArnold How can we help? Do you need acces to an AHM32 for testing purposes? We have one available at the moment.

JonasArnold commented:
> @JonasArnold How can we help? Do you need acces to an AHM32 for testing purposes? We have one available at the moment.

I do have access to an AHM32 aswell. The issue is that I cannot get the update script to run, so I need somebody with more experience with bitfocus companion module development to help me out with the code or debugging it.

Unfortunatly I think you will have to wait until the update script runs and the code is merged.

bryce-seifert commented:
Hi, I will take a look and provide help later today

bryce-seifert commented:
@JonasArnold I just tested this, and all seems to be working fine for me! In order to test, I used an old version of Companion to create a page of buttons with the actions to test. Then I exported that page, and imported the page on my dev version of Companion. I saw the logging for the upgrades, and verified all the action values weren't undefined.

So I think this should be good to go!

JonasArnold commented:
> @JonasArnold I just tested this, and all seems to be working fine for me! In order to test, I used an old version of Companion to create a page of buttons with the actions to test. Then I exported that page, and imported the page on my dev version of Companion. I saw the logging for the upgrades, and verified all the action values weren't undefined.
> 
> So I think this should be good to go! Let me know if you have anything else you want to do before I merge

Thanks a lot for testing!
Using your method I was able to test the update script aswell and I even found a little bug :) If the action.options.inputChannel had the value of 0 it was not detected, which lead to it not being updated. I fixed this now and I think it would be ready to be merged now. 
I already bumped the version to 2.1.0 in the last commit.

Thanks a lot for your support @bryce-seifert !
I'm happy to work on the module in the future if there are any more enhancements required.

By the way, this is how I was testing the update script: I disabled the development module => Companion took the original version of the module, added some actions and stuff, then re-enabled the development module => Companion takes the development module. I suspected it would use the update script in that case, but I'm probably wrong...

bryce-seifert commented:
Re: upgrade script testing, yeah, there used to be a more explicit was to test the upgrade scripts, but I don't think that's existed since 3.0. It probably needs some work on the core end.

I'm going to go ahead and merge and release this!

bryce-seifert commented:
@fenderwall These features form this PR should now be available for testing in the latest Companion beta

bryce-seifert commented:
@JonasArnold Thanks very much for your work on this! Also, I would recommend joining the [Slack](https://bitfocus.io/api/slackinvite) if you're open to it for easier communication. I know there are some open requests for this module that I just haven't had time to get to, so if any interest / are of use to you I'm sure the users would be happy.

fenderwall commented:
Hi! 

I’ve discovered a bug in the AHM Companion module. I tested this both in the previous version and the updated one, and the issue persists in both.

I’m using the AHM Matrix as part of my GreenGo intercom setup, with multiple microphones connected to the inputs, and speakers and 4-wire interfaces connected to the outputs. I create various routes using zones. For example, Mic 1 is routed to a speaker on Output 6.

In Companion, I’ve created a button with two actions. First, I mute Input 1 (Mic 1) to Zone 6 using the “Mute input to Zone” function. In the second action, I unmute the same input to the same zone. For feedback, I use the “Change background when input to zone is muted” option, with Input 1 and Zone 6 selected. This works perfectly when using just one button.

However, when I add a second button with the same input (Input 1) but for a different zone (e.g., Zone 7), including mute, unmute, and feedback actions for this zone, this second button interferes with the feedback of the first button. The second button doesn’t change the mute functions of the first button on the AHM32, but it sets the feedback of the first button to “unmute.”

I hope you can assist me with this issue. I had done some testing and anticipated that the AHM32 would integrate smoothly. However, this bug is making it challenging to use the setup in our production environment.

Thank you in advance, and let me know if I can provide any assistance.
